accepted/2024/NuGet-Audit-for-PackageDownload.md

```diff
@@ -0,0 +1,123 @@
+To prevent user confusion about SDK-included dependencies, we are introducing a new attribute, `AddedBy`, to identify the origin of each `<PackageDownload>`.
+This will provide users with clear, actionable security information while helping them distinguish between SDK-provisioned packages and those they added manually.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+
+We are expanding NuGet Audit to `<PackageDownload>` packages to improve security notifications, addressing gaps in vulnerability alerts that currently miss these packages.
+This change prevents confusion by clarifying SDK-managed PackageDownloads in the warning reports, ensuring users have an actionable warning.
+The expected outcome is a more precise and useful vulnerability reporting experience, enabling users to address actual risks without confusing SDK-related alerts.
+
+## Explanation
+
+### Functional explanation
+
+The NuGet Audit now supports `<PackageDownload>` packages, allowing it to detect and alert users about vulnerabilities in any user-specified dependencies downloaded via `<PackageDownload>`. 
+This means you’ll now receive warnings if a package added via `<PackageDownload>` has known security issues, so you can quickly address them.
+
+To prevent confusion about SDK-managed packages, we introduce a new attribute, `AddedBy`, on `<PackageDownload>` entries.
+This attribute will specify the source of the package, with values like `"dotnetsdk"` for SDK-managed packages and `"user"` for user-added dependencies.
+When a package marked with `AddedBy="dotnetsdk"` has a vulnerability, the warning will indicate that the package was added by the SDK.
+This will help users understand that the package originates from the SDK and is not something they explicitly added.
+
+| Attribute | Value     | Warning   |
+|-----------|-----------|-----------|
+| AddedBy   | dotnetsdk | warning NU1906: Package 'System.Text.Json' 1.0.0, added by '.NET SDK', has a known vulnerability. Refer to the vulnerability details at https://github.com/advisories/GHSA-g3q9-xf95-8hp5 to assess any necessary actions. |
+| AddedBy   | user      | warning NU1906: Package 'System.Text.Json' 1.0.0, added by 'user', has a known vulnerability. Refer to the vulnerability details at https://github.com/advisories/GHSA-g3q9-xf95-8hp5 to assess any necessary actions. |
+| AddedBy   | undefined | warning NU1906: Package 'System.Text.Json' 1.0.0 has a known vulnerability. Refer to the vulnerability details at https://github.com/advisories/GHSA-g3q9-xf95-8hp5 to assess any necessary actions. |
+
+However, all of these warnings will be gated by the `SDKAnalysisLevel` property.
+The warnings will only be shown if the user is using a .NET SDK version that introduced these warnings or newer.
+For example, if these warnings are introduced in SDK version 9.0.300, the warnings will only appear if `SDKAnalysisLevel` is defined and set to 9.0.300 or higher.
```

Comment:
What we learned from the .NET 9.0.100 GA release is that `SDKAnalysisLevel` does not meet our customers or leaderships expectations. So, we can't use it as the gate.

I think we need to add a new config `<NuGetAuditPackageDownloads>`, which for now will default to `false`.

Once we have a plan for changing `NuGetAuditMode`'s defaults, we'll use the same plan to change this new config setting as well.

Reply:
Per offline chat, tying packagedownload audit to level all is probably reasonable even though these are direct references. That way we're not opting people in earlier than they expect.

Reply:
After considering the options, I decided to go with adding the `direct-with-packagedownload` value to `NuGetAuditMode` instead of introducing a separate `<NuGetAuditPackageDownloads>` configuration. This approach keeps audit-related settings centralized, avoids introducing new properties